src/main/java/decodes/db/DecodesScript.java

```diff
@@ -16,538 +16,721 @@
+     */
+    public String getHeaderType()
+    {
+        int idx = scriptType.indexOf(':');
+        if (idx < 0 || scriptType.length() <= idx+1)
+        {
+            return null;
+        }
+        return scriptType.substring(idx+1);
+    }
+
+    public void addMissing(String symbol)
+    {
+        missingSymbols.add(symbol);
+    }
+
+    public boolean isMissingSymbol(String symbol)
+    {
+        return missingSymbols.contains(symbol);
+    }
+
+    public void setIncludePMs(ArrayList<String> includePMs)
+    {
+        this.includePMs = includePMs;
+        if (includePMs != null && includePMs.size() == 0)
+            this.includePMs = null;
+        if (this.includePMs != null)
+        {
+            logger.debug1("setIncludePMs: includePMs has "
+                         + includePMs.size() + " strings. [0]=" + includePMs.get(0));
+        }
+    }
+
+    /**
+     * Add the PMs as sensor values.
+     */
+    private void addPMs(DecodedMessage decodedMessage)
+    {
+        RawMessage rm = decodedMessage.getRawMessage();
+        if (rm == null)
+        {
+            return;
+        }
+        Platform p = null;
+        try
+        {
+            p = rm.getPlatform();
+        }
+        catch(UnknownPlatformException ex)
+        {
+            return;
+        }
+        if (p == null)
+        {
+            return;
+        }
+        PlatformConfig pc = p.getConfig();
+        if (pc == null)
+        {
+            return;
+        }
+      nextPM:
+        for(String pmname : includePMs)
+        {
+            logger.debug1("addPMs looking for '" + pmname + "'");
+            Variable v = rm.getPM(pmname);
+            if (v == null)
+            {
+                logger.info(
+                    "addPMs: Message for '" + new String(rm.getHeader())
+                    + "' does not have requested performance measurement '" + pmname + "' -- ignored.");
+                continue;
+            }
+            for (ConfigSensor cs : pc.getSensorVec())
+            {
+                if (cs.sensorName.equalsIgnoreCase(pmname))
+                {
+                    decodedMessage.addSample(cs.sensorNumber, v, 0);
+                    logger.debug1("addPM: added pm '" + pmname + "' to sensor " + cs.sensorNumber + " with value '" + v.toString() + "'");
+                    continue nextPM;
+                }
+            }
+            // Fell through, no matching sensor:
+            logger.info("addPMs: The platform config for '" + new String(rm.getHeader())
+                + "' does not have a sensor named '" + pmname + "' -- discarded.");
+        }
+    }
+
+    /**
+     * Start a builder for a decodes script with a given reader.
+     * @param scriptReader concrete implementation that provides the required statements
+     * @return a Builder object for further operations.
+     * @since 2022-11-05
+     */
+    public static DecodesScriptBuilder from(final DecodesScriptReader scriptReader)
```

Comment:
`final` is unnecessary > java 7

Reply:
Thanks, apparently I mixed that up with c++'s "const" keyword. Very different meaning.